src/requests.rs

```diff
@@ -423,16 +410,96 @@ macro_rules! request {
-request!(SignatureUploadRequest from OriginalSignatureUploadRequest extracts signed_keys: json);
 request!(RoomMessageRequest from OriginalRoomMessageRequest extracts room_id: string, txn_id: string, event_type: event_type, content: json);
 request!(KeysBackupRequest from OriginalKeysBackupRequest extracts version: string and groups rooms);
 
-// Other Requests
-request!(SigningKeysUploadRequest from OriginalUploadSigningKeysRequest groups master_key, self_signing_key, user_signing_key);
+// Specific conversion for signature upload as they can be returned directly by
+// some verification API and not via outgoing requests.
+impl TryFrom<&OriginalSignatureUploadRequest> for SignatureUploadRequest {
+    type Error = serde_json::Error;
+    fn try_from(request: &OriginalSignatureUploadRequest) -> Result<Self, Self::Error> {
+        {
+            Ok(SignatureUploadRequest {
+                id: None,
+                signed_keys: serde_json::to_string(&request.signed_keys)?.into(),
+            })
+        }
+    }
+}
+
+// Not handled by macro as there is no request_id.
+impl TryFrom<&OriginalUploadSigningKeysRequest> for SigningKeysUploadRequest {
+    type Error = serde_json::Error;
+    fn try_from(request: &OriginalUploadSigningKeysRequest) -> Result<Self, Self::Error> {
+        {
+            Ok(SigningKeysUploadRequest {
+                body: {
+                    let mut map = serde_json::Map::new();
+                    map.insert(
+                        "master_key".to_owned(),
+                        serde_json::to_value(&request.master_key).unwrap(),
+                    );
+                    map.insert(
+                        "self_signing_key".to_owned(),
+                        serde_json::to_value(&request.self_signing_key).unwrap(),
+                    );
+                    map.insert(
+                        "user_signing_key".to_owned(),
+                        serde_json::to_value(&request.user_signing_key).unwrap(),
+                    );
+                    let object = serde_json::Value::Object(map);
+                    serde_json::to_string(&object)?.into()
+                },
+            })
+        }
+    }
+}
+
+// Manual into/from for signature upload requests as the id is optional.
+impl TryFrom<(String, &OriginalSignatureUploadRequest)> for SignatureUploadRequest {
+    type Error = serde_json::Error;
+    fn try_from(
+        (request_id, request): (String, &OriginalSignatureUploadRequest),
+    ) -> Result<Self, Self::Error> {
+        {
+            Ok(SignatureUploadRequest {
+                id: Some(request_id.into()),
+                signed_keys: serde_json::to_string(&request.signed_keys)?.into(),
+            })
+        }
+    }
+}
+
+// Specific conversion for to-device as they can be returned directly instead of
+// via outgoing requests. See `share_room_key`, the request_id is the txn_id.
```

Comment:
```suggestion
// ToDeviceRequests can be returned directly from `OlmMachine::share_room_key`, instead of
// being wrapped in a `matrix_sdk_crypto::OutgoingRequest` via `OlmMachine::outgoing_requests`.
// 
// We therefore need a custom conversion implementation to deal with that case.
```